Recognise cygwin platform and open urls accordingly

TDaglis opened the pull request:
Currently, click does not recognise when it is run inside a [cygwin](https://www.cygwin.com/) terminal.
This PR fixes [launch](http://click.pocoo.org/5/utils/#launching-applications). Instead of doing nothing when run inside a cygwin terminal, it properly opens the default application (picture viewer, browser, etc)

I did not find a way to make `locate` option work though so I just let it pass to the default case (where nothing happens).

untitaker commented:
According to http://stackoverflow.com/a/577609 we could also just replace `start` with `cmd /c start` in the Windows section for proper behavior within cygwin.

TDaglis commented:
:+1: will test and update accordingly

TDaglis commented:
@untitaker I updated it to `cmd /c start` and merged Cygwin case with Windows case.
The only issue I faced was the following:
With `cygstart`, to open a file via its absolute path, the "cygwin" path was used
With `cmd /c start`, to open a file by providing its absolute path, its "Windows" path has to be used.
For example, to open `test.png` in `C:\\cygwin64\\home\\Thanasis\\mydir` I'll have to:
- `cygstart "/home/Thanasis/mydir/test.png"` or
- `cmd /c start "C:\cygwin64\home\Thanasis\mydir\test.png"`

I'm not sure which should be considered correct. `os.path.realpath('test.png')` run in a cygwin terminal would produce the first path.

TDaglis commented:
I also just noticed that attempting to open files in any cygwin subdirectory with `cmd /c start` results to failure with `Access is denied.`, while `cygstart` opens files no matter the location.
I have no idea why this happens... I'll have to  look into it

untitaker commented:
I see, in that case the previous version (with the special conditional) is probably better, although the codepath for `locate` is problematic.

TDaglis commented:
Reverted to previous version.
Also updated codepath for `locate`.
Now, when `locate=True`, instead of opening the url, the parent directory opens. What I could not achieve was selecting the file in the new the File Explorer.

untitaker commented:
Excellent, thanks!
